lib/src/components/NxScrollRender/NxScrollRender.tsx

```diff
@@ -0,0 +1,133 @@
+
+const DEFAULT_INITIAL_CHILD_COUNT = 40;
+
+export default function NxScrollRender({ children, reuseChildren, initialChildCount, spacerEl }: Props) {
+  const fullParent = children,
+
+      childArray = useMemo(() => React.Children.toArray(fullParent.props.children) as ReactElement[],
+          [fullParent.props.children]),
+
+      childCount = childArray.length,
+
+      parentRef = useRef<HTMLElement>(null),
+      leadingSpacerRef = useRef<HTMLElement | null>(null),
+      leadingSpacerRefFn = (e: HTMLElement | null) => { leadingSpacerRef.current = e; },
+      trailingSpacerRef = useRef<HTMLElement | null>(null),
+      trailingSpacerRefFn = (e: HTMLElement | null) => { trailingSpacerRef.current = e; },
+
+      [parentHeight, setParentHeight] = useState<number | null>(null),
+      [childHeight, setChildHeight] = useState<number | null>(null),
+      normalizedChildHeight = childHeight ?? 0,
+      renderedChildCount = parentHeight == null || childHeight == null ?
+        null : Math.min(Math.ceil(divOrZero(parentHeight, childHeight)) + 2, childCount),
+
+      cloneWithKey = (child: ReactElement, idx: number) =>
+        React.cloneElement(child, { key: idx % (renderedChildCount as number) }),
+      keyedChildren = useMemo(
+          () => reuseChildren === false || isNil(renderedChildCount) ? childArray : childArray.map(cloneWithKey),
+          [childArray, reuseChildren, renderedChildCount]
+      ),
+
+      sumChildHeight = normalizedChildHeight * childCount,
+      renderedChildHeight = renderedChildCount == null ? 0 : normalizedChildHeight * renderedChildCount,
+
+      [firstRenderedChildIdx, setFirstRenderedChildIdx] = useState(0),
+
+      renderedChildCountWithDefault = renderedChildCount ?? initialChildCount ?? DEFAULT_INITIAL_CHILD_COUNT,
+      renderedRealChildren =
+          keyedChildren.slice(firstRenderedChildIdx, firstRenderedChildIdx + renderedChildCountWithDefault),
+
+      sumSpacerHeight = sumChildHeight - renderedChildHeight,
+      clampSpacerHeight = clamp(0, sumSpacerHeight),
+      leadingSpacerHeight = clampSpacerHeight(firstRenderedChildIdx * normalizedChildHeight),
+      trailingSpacerHeight = renderedChildCount == null ? 0 : clampSpacerHeight(
+          sumChildHeight - (leadingSpacerHeight + renderedChildCount * normalizedChildHeight)
+      );
+
+  const renderedChildren = childCount ? (
+    <>
+      {React.createElement(spacerEl ?? 'div', {
+        ref: leadingSpacerRefFn,
+        role: 'presentation',
+        className: 'nx-scroll-render__spacer',
+        style: { height: leadingSpacerHeight }
+      })}
+      {renderedRealChildren}
+      {React.createElement(spacerEl ?? 'div', {
+        ref: trailingSpacerRefFn,
+        role: 'presentation',
+        className: 'nx-scroll-render__spacer',
+        style: { height: trailingSpacerHeight }
+      })}
+    </>
+  ) : null;
+
+  const updateRendering = useThrottleCallback(useCallback(function updateRendering() {
```

Comment:
I'm just trying to wrap my head around this line with its multiple callbacks. Would you be able to clarify for me?

Reply:
> I'm just trying to wrap my head around this line with its multiple callbacks. Would you be able to clarify for me?

It's probably best explained by the [`useThrottleCallback` documentation](https://www.npmjs.com/package/@react-hook/throttle).  `useThrottleCallback` is described as
> A hook that invokes its callback at most X frames per second.

In the description of its first parameter, it notes
> This is the callback you want to throttle. You need to wrap closures/unstable callbacks in useCallback() so that they are stable, otherwise throttling will break between renders.

So the goal is to create a function that can be called multiple times, potentially across different renders of this component, but which will only execute with a given maximum frequency even if it is called more often than that maximum.  In order for `useThrottleCallback` to always return the same function instance, which is necessary to maintain that call-frequency management, it must always be given the same underlying function instance.  `useCallback` ensures this.  Without `useCallback`, each call of the `NxScrollRender` render function would pass a new function instance to `useThrottleCallback`, and so `useThrottleCallback` would return a new instance, effectively resetting the throttling timers.  `useCallback` ensures that instead of passing a new instance each time, we always pass the same instance, except when the dependencies declared in `useCallback`'s second parameter change.

Reply:
> So the goal is to create a function that can be called multiple times, potentially across different renders of this component, but which will only execute with a given maximum frequency even if it is called more often than that maximum. In order for `useThrottleCallback` to always return the same function instance, which is necessary to maintain that call-frequency management, it must always be given the same underlying function instance. `useCallback` ensures this. Without `useCallback`, each call of the `NxScrollRender` render function would pass a new function instance to `useThrottleCallback`, and so `useThrottleCallback` would return a new instance, effectively resetting the throttling timers. `useCallback` ensures that instead of passing a new instance each time, we always pass the same instance, except when the dependencies declared in `useCallback`'s second parameter change.

Thanks for this explanation, it helps a lot!